## Análise pós-clusterização
***

In [156]:
import pandas as pd
import numpy as np

In [249]:
path = '/Users/felipebarreto/Desktop/TCC/Bases geradas/'
df1 = pd.read_csv('/Users/felipebarreto/Desktop/dados_acidentes_tratados_completo.csv', sep = ',')   ## dados originais tratados
df2 = pd.read_csv(path + 'centroides_padronizados.csv', sep = ';')   ## centróides de cada cluster     //
df3 = pd.read_csv(path + 'variaveis_quanti_agg.csv', sep = ';')      ## igual ao df4, sem padronização // 
df4 = pd.read_csv(path + 'variaveis_quanti_pad_agg.csv', sep = ';')  ## igual ao df3, com padronização // 
df5 = pd.read_csv(path + 'medias_var_cluster.csv', sep = ';', decimal= '.')        ## médias de cada cluster         // 
df6 = pd.read_csv(path + 'clusters_7.csv', sep = ';') # var quantitativas completas, nem todas foram utilizadas para a clusterização

In [245]:
df1.groupby('tracado_via_tipo_1')['tracado_via_tipo_1'].count().keys()

Index([False, True], dtype='object', name='tracado_via_tipo_1')

In [236]:
pd.DataFrame(df1.tracado_via_tipo_1.value_counts(normalize=True)).reset_index()

index  tracado_via_tipo_1
0   True            0.707132
1  False            0.292868

In [254]:
df6

municipio  pessoas  mortos  feridos_leves  feridos_graves  ilesos  \
0            curitiba    53273     581          10382            2872   36629   
1            sao jose    44027     191           8828            1928   31326   
2           guarulhos    40350     430           8282            1040   27409   
3               betim    39527     329           6840            1718   29551   
4     duque de caxias    37859     407           6388            1273   28066   
...               ...      ...     ...            ...             ...     ...   
2258           viamao        2       0              0               0       2   
2259       vila pavao        1       0              0               0       1   
2260     virginopolis        2       0              1               0       1   
2261  vitorino freire        2       0              0               0       2   
2262        xangri-la        1       0              0               0       1   

      ignorados  feridos  veiculos   freq  share_tracado_via_tipo_1  \
0          2808    13254     46614  23714                  0.834781   
1          1752    10756     40653  20102                  0.784648   
2          3188     9322     36372  18109                  0.942239   
3          1085     8558     35159  17601                  0.784103   
4          1723     7661     32465  17531                  0.842393   
...         ...      ...       ...    ...                       ...   
2258          0        0         2      1                  1.000000   
2259          0        0         1      1                  1.000000   
2260          0        1         1      1                  0.000000   
2261          0        0         2      1                  1.000000   
2262          0        0         1      1                  0.000000   

      share_tracado_via_tipo_2  share_is_night  share_is_weekend  \
0                     0.155604        0.375221          0.438306   
1                     0.205552        0.369018          0.414785   
2                     0.052018        0.410293          0.422055   
3                     0.208625        0.361911          0.489234   
4                     0.153785        0.380811          0.430152   
...                        ...             ...               ...   
2258                  0.000000        0.000000          0.000000   
2259                  0.000000        0.000000          0.000000   
2260                  1.000000        0.000000          1.000000   
2261                  0.000000        0.000000          0.000000   
2262                  1.000000        1.000000          0.000000   

      share_is_single_lane  cluster  
0                 0.106477        1  
1                 0.043379        1  
2                 0.055221        1  
3                 0.048804        1  
4                 0.195710        1  
...                    ...      ...  
2258              0.000000        3  
2259              1.000000        2  
2260              1.000000        4  
2261              1.000000        2  
2262              1.000000        4  

[2263 rows x 16 columns]

In [288]:
def _status_qtd(var):
    a = pd.merge(df1, df6[['municipio','cluster']], how='left')\
          .groupby(['cluster',var])[['id']]\
          .count()\
          .reset_index()\
          .rename(columns={var:'status','id':var})\
          .set_index(['cluster','status'])
    return a

In [293]:
_status_qtd('tracado_via_tipo_1')\
    .join(_status_qtd('is_weekend'))\
    .join(_status_qtd('is_night'))\
    .join(_status_qtd('is_single_lane'))

tracado_via_tipo_1  is_weekend  is_night  is_single_lane
cluster status                                                          
1       False                62988      175841    199847          247456
        True                249718      136865    112859           65250
2       False                80012      175564    200865           20302
        True                263650      168098    142797          323360
3       False                65031      114960    131837          185778
        True                153379      103450     86573           32632
4       False                86222       78425     95846            7071
        True                 66154       73951     56530          145305
5       False               214713      379447    440833          368479
        True                496002      331268    269882          342236

In [285]:
pd.join(_status_qtd('is_night'),_status_qtd('is_weekend'))

AttributeError: module 'pandas' has no attribute 'join'

In [283]:
_status_qtd('is_night')

cluster  status  is_night
0        1   False    199847
1        1    True    112859
2        2   False    200865
3        2    True    142797
4        3   False    131837
5        3    True     86573
6        4   False     95846
7        4    True     56530
8        5   False    440833
9        5    True    269882

In [282]:
pd.merge(
    _status_qtd('tracado_via_tipo_1'),_status_qtd('is_weekend'),_status_qtd('is_night')
)

ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [280]:
_status_qtd('tracado_via_tipo_1')\
    .pd.merge(_status_qtd('is_weekend'), how=['cluster','status'])

AttributeError: 'DataFrame' object has no attribute 'pd'

In [273]:
pd.merge(df1, df6[['municipio','cluster']], how='left')\
    .groupby(['cluster','tracado_via_tipo_1'])[['id']]\
    .count()\
    .reset_index()\
    .rename(columns={'tracado_via_tipo_1':'status','id':'tracado_via_tipo_1'})

cluster  status  tracado_via_tipo_1
0        1   False               62988
1        1    True              249718
2        2   False               80012
3        2    True              263650
4        3   False               65031
5        3    True              153379
6        4   False               86222
7        4    True               66154
8        5   False              214713
9        5    True              496002

In [252]:
df6.groupby(['cluster','share_tracado_via_tipo_1'])[['cluster']].count()

cluster
cluster share_tracado_via_tipo_1         
1       0.632138                        1
        0.669109                        1
        0.680138                        1
        0.694181                        1
        0.706108                        1
...                                   ...
5       0.960695                        1
        0.963798                        1
        0.967316                        1
        0.971684                        1
        0.972609                        1

[1602 rows x 1 columns]

In [158]:
display(
    df2.style.background_gradient(cmap='Greens').set_precision(2),
    df5.style.background_gradient(cmap='Reds').set_precision(2)
    )

### Análise

In [166]:
df = df3.drop(['pessoas','ignorados','feridos','veiculos'], axis=1)[[
    'municipio','cluster','freq','mortos','feridos_graves','feridos_leves','ilesos','share_tracado_via_tipo_1',
    'share_tracado_via_tipo_2','share_is_night','share_is_weekend','share_is_single_lane'
]].drop_duplicates(subset = ['municipio'])


## Separação dos clusters (c1,c2,c3,c4,c5,c6,c7)
n = len(df.cluster.drop_duplicates())
for i in range(1,n + 1):
    globals()['c' + str(i)] = df[df.cluster == i].drop(['cluster'], axis=1)

df.head(5).style.background_gradient(cmap='Blues').set_precision(2)

In [204]:
c1.mortos.sum()

7928

In [205]:
c1.mortos.mean()

330.3333333333333

In [160]:
def _loc(df, cluster, col):
    return(round(df[df.cluster == 'Cluster '+str(cluster)][col].iloc[0],2))

In [171]:
round(df.mortos.mean(),2)

39.59

In [193]:
globals()['mortos_c' + str(1)] = 2

In [219]:
## Tamanho dos clusters // Quantidade de municípios
print(
    f""">> DADOS GERAIS - Acidentes de 2007 a 2021 (16 anos)<<
    - Total de municípios analisados: {len(df)}
    - Mortos = {df.mortos.sum()} | Média = {round(df.mortos.mean(),2)}
    - Feridos graves = {df.feridos_graves.sum()} | Média = {round(df.feridos_graves.mean(),2)}
    - Feridos leves = {df.feridos_leves.sum()} | Média = {round(df.feridos_leves.mean(),2)}
    - Ilesos = {df.ilesos.sum()} | Média = {round(df.ilesos.mean(),2)}
    """
)

for i in range(1,n+1):

    print(f"""Cluster {i}:
        - {len(globals()['c' + str(i)])} municípios --> {round(100*df[df.cluster == i].shape[0] / df.shape[0],2)}% do total
        - Mortos = {globals()['c' + str(i)].mortos.sum()},  {round(100*globals()['c' + str(i)].mortos.sum() / df.mortos.sum(),2)}% do total | Média = {_loc(df5, i, col='mortos')}, {round(100*_loc(df5, i, col='mortos') / df.mortos.mean(),2)}% da média geral
        - Feridos graves = {globals()['c' + str(i)].feridos_graves.sum()},  {round(100*globals()['c' + str(i)].feridos_graves.sum() / df.feridos_graves.sum(),2)}% do total | Média = {_loc(df5, i, col='feridos_graves')}, {round(100*_loc(df5, i, col='feridos_graves') / df.feridos_graves.mean(),2)}% da média geral
        - Feridos Leves = {globals()['c' + str(i)].feridos_leves.sum()},  {round(100*globals()['c' + str(i)].feridos_leves.sum() / df.feridos_leves.sum(),2)}% do total | Média = {_loc(df5, i, col='feridos_leves')}, {round(100*_loc(df5, i, col='feridos_leves') / df.feridos_leves.mean(),2)}% da média geral
        - Ilesos = {globals()['c' + str(i)].ilesos.sum()},  {round(100*globals()['c' + str(i)].ilesos.sum() / df.ilesos.sum(),2)}% do total | Média = {_loc(df5, i, col='ilesos')}, {round(100*_loc(df5, i, col='ilesos') / df.ilesos.mean(),2)}% da média geral
        - Média do share de traçado via tipo 1 = {_loc(df5, i, col='share_tracado_via_tipo_1')}
        - Média do share de é noite = {_loc(df5, i, col='share_is_night')}
        - Média do share de é final de semana = {_loc(df5, i, col='share_is_weekend')}
        - Média do share de é pista simples = {_loc(df5, i, col='share_is_single_lane')}
    """)

>> DADOS GERAIS - Acidentes de 2007 a 2021 (16 anos)<<
    - Total de municípios analisados: 2263
    - Mortos = 89587 | Média = 39.59
    - Feridos graves = 297414 | Média = 131.42
    - Feridos leves = 837677 | Média = 370.16
    - Ilesos = 2395201 | Média = 1058.42
    
Cluster 1:
        - 24 municípios --> 1.06% do total
        - Mortos = 7928,  8.85% do total | Média = 330.33, 834.43% da média geral
        - Feridos graves = 42051,  14.14% do total | Média = 1752.12, 1333.17% da média geral
        - Feridos Leves = 130439,  15.57% do total | Média = 5434.96, 1468.26% da média geral
        - Ilesos = 482400,  20.14% do total | Média = 20100.0, 1899.06% da média geral
        - Média do share de traçado via tipo 1 = 0.8
        - Média do share de é noite = 0.36
        - Média do share de é final de semana = 0.44
        - Média do share de é pista simples = 0.24
    
Cluster 2:
        - 1264 municípios --> 55.86% do total
        - Mortos = 34174,  38.15% do total | Média = 

In [12]:
#df5 = df5.set_index('cluster')

df5.style.background_gradient(cmap='Greens')\
    #.background_gradient(cmap='coolwarm', axis=None, vmin=0, vmax=1)
    #.highlight_null(null_color='#f1f1f1')  # Color NaNs grey
    .set_precision(2)

#### Observações:

- Total de Municípios analisados: 2263
- Municípios por região:
    - N ---> 
    - NE --> 
    - CO --> 
    - S ---> 
    - SE -->  
- Municípios por UF:
    - sdc

## Cluster 1

In [105]:
df5[df5.cluster == 'Cluster 5'].feridos_leves.iloc[0]

1689.63076923077

In [101]:
c1.shape

(493, 10)

In [ ]:
c1.groupby('uf')[['freq']].sum().sort_values(by = ['freq'], ascending = False)

freq
uf       
mg  33232
sc  28153
rs  21568
ba  13154
pr  12833
es   6291
al   4140
rj   1772
go   1314
sp   1253
pe    770
ma    712
pi    587
ce    482
se    323
rn    297
pa    239
pb    179
ap    147
ro    103
to     67
ac     33
ms     15
am      3
mt      2

In [ ]:
c1.head(50)

uf              municipio  freq  mortos  feridos_graves  feridos_leves  \
312  sc           aguas mornas  1520      77             331            906   
331  mg    bela vista de minas  1408      62             366            930   
346  sc   sao cristovao do sul  1372      71             209            682   
350  mg           campos altos  1357      83             219            796   
354  sc        rancho queimado  1335      66             260            742   
366  es        rio novo do sul  1255      81             243            579   
372  sc                  xaxim  1224      63             200            681   
377  sc         alfredo wagner  1201      70             251            816   
379  pr            pato branco  1196      32             190            666   
380  sc             ponte alta  1194      60             193            586   
386  rs        cacapava do sul  1174      51             158            459   
391  rj                 parati  1171      57             219            597   
400  rs              cruz alta  1150      52             144            508   
410  sc          santa cecilia  1124      70             220            564   
417  mg             nova uniao  1094      66             283            529   
419  mg          corrego danta  1092      76             235            597   
420  es                 iconha  1090      42             181            499   
421  mg                   ibia  1082      74             192            644   
434  rs            dois irmaos  1041      25             167            575   
444  sc                joacaba  1015      46             154            561   
445  es               anchieta  1008      93             186            497   
446  sc               anchieta  1008      93             186            497   
451  mg             pouso alto  1004      33             116            580   
464  rs                   ijui   975      36             136            521   
478  ba                arataca   934      71             123            408   
482  sc                  irani   930      40             178            553   
483  sc          vargem bonita   929      64             188            515   
490  mg  sao domingos do prata   915      81             221            482   
493  rs                   bage   904      19              99            337   
506  ba                   catu   878      51             178            487   
508  sc            curitibanos   860      62             166            564   
511  mg        pocos de caldas   855      38             177            522   
512  pr       nova laranjeiras   854      86             137            404   
517  rs       marques de souza   844      36             121            430   
523  go               mineiros   832      81             168            466   
526  mg              rio casca   830      34             172            411   
533  rs             pouso novo   816      52             102            370   
534  rs                panambi   815      57             124            476   
535  sc              concordia   814      42             188            442   
538  ba                itapebi   806      90             118            405   
543  rs         lagoa vermelha   799      37             132            447   
551  pr      cornelio procopio   793      32             150            505   
553  mg           alem paraiba   790      66             150            412   
555  ba    sao jose da vitoria   788      65             140            393   
567  es             itapemirim   775      38             114            310   
569  al                 satuba   774      43             247            492   
570  mg                inhapim   773      81             241            481   
572  sc         bocaina do sul   767      53             163            472   
573  sc                 corupa   764      35             207            531   
574  ba           ibirapitanga   763     119           